app/src/organisms/ProtocolUpload/hooks/useCurrentProtocol.ts

```diff
@@ -0,0 +1,38 @@
+import * as React from 'react'
+import last from 'lodash/last'
+import { useProtocolQuery } from '@opentrons/react-api-client'
+import { useCurrentRun } from './useCurrentRun'
+
+import type { Protocol } from '@opentrons/api-client'
+
+export function useCurrentProtocol(): Protocol | null {
+  const currentRun = useCurrentRun()
+
+  const enableProtocolPolling = React.useRef<boolean>(
+    currentRun?.data.protocolId != null
+  )
+  const { data: protocolRecord } = useProtocolQuery(
+    (currentRun?.data?.protocolId as string) ?? null,
+    { staleTime: Infinity },
+    enableProtocolPolling.current
+  )
+
+  const mostRecentAnalysis = last(protocolRecord?.data.analyses) ?? null
+
+  React.useEffect(() => {
+    if (
+      mostRecentAnalysis?.status === 'completed' &&
+      enableProtocolPolling.current
+    ) {
+      enableProtocolPolling.current = false
+    } else if (
+      currentRun?.data.protocolId != null &&
+      !enableProtocolPolling.current &&
+      mostRecentAnalysis?.status !== 'completed'
+    ) {
+      enableProtocolPolling.current = true
+    }
+  }, [mostRecentAnalysis?.status, currentRun?.data.protocolId])
```

Comment:
I think this effect can be simplified:

- Checking `enableProtocolPolling` doesn't do anything, because setting `false` to `false` or `true` to `true` is harmless
- Checking that `currentRun?.data.protocolId` is defined doesn't do anything because if it isn't defined then `mostRecentAnalysis` won't be defined
- It's ok to set `enableProtocolPolling` to `true` even if `mostRecentAnalysis` is undefined for the same reason as my comment above

```suggestion
  React.useEffect(() => {
    if (mostRecentAnalysis?.status === 'completed') {
      enableProtocolPolling.current = false
    } else {
      enableProtocolPolling.current = true
    }
  }, [mostRecentAnalysis?.status])
```

We may also want to consider resolving these "uncovered" warnings.